pkg/validation/components_test.go

```diff
@@ -0,0 +1,253 @@
+		ComponentUnion: v1alpha2.ComponentUnion{
+			Container: &v1alpha2.ContainerComponent{
+				Container: v1alpha2.Container{
+					Image:        image,
+					Env:          envs,
+					VolumeMounts: volMounts,
+					MountSources: &mountSources,
+				},
+				Endpoints: endpoints,
+			}}}
+}
+
+// generateDummyVolumeComponent returns a dummy volume component for testing
+func generateDummyVolumeComponent(name, size string) v1alpha2.Component {
+
+	return v1alpha2.Component{
+
+		Name: name,
+		ComponentUnion: v1alpha2.ComponentUnion{
+			Volume: &v1alpha2.VolumeComponent{
+				Volume: v1alpha2.Volume{
+					Size: size,
+				},
+			},
+		},
+	}
+}
+
+// generateDummyOpenshiftComponent returns a dummy Openshift component for testing
+func generateDummyOpenshiftComponent(name string, endpoints []v1alpha2.Endpoint, uri string) v1alpha2.Component {
+
+	return v1alpha2.Component{
+		Name: name,
+		ComponentUnion: v1alpha2.ComponentUnion{
+			Openshift: &v1alpha2.OpenshiftComponent{
+				K8sLikeComponent: v1alpha2.K8sLikeComponent{
+					K8sLikeComponentLocation: v1alpha2.K8sLikeComponentLocation{
+						Uri: uri,
+					},
+					Endpoints: endpoints,
+				},
+			},
+		},
+	}
+}
+
+// generateDummyKubernetesComponent returns a dummy Kubernetes component for testing
+func generateDummyKubernetesComponent(name string, endpoints []v1alpha2.Endpoint, uri string) v1alpha2.Component {
+
+	return v1alpha2.Component{
+		Name: name,
+		ComponentUnion: v1alpha2.ComponentUnion{
+			Kubernetes: &v1alpha2.KubernetesComponent{
+				K8sLikeComponent: v1alpha2.K8sLikeComponent{
+					K8sLikeComponentLocation: v1alpha2.K8sLikeComponentLocation{
+						Uri: uri,
+					},
+					Endpoints: endpoints,
+				},
+			},
+		},
+	}
+}
+
+// generateDummyPluginComponent returns a dummy Plugin component for testing
+func generateDummyPluginComponent(name, url string) v1alpha2.Component {
+
+	return v1alpha2.Component{
+		Name: name,
+		ComponentUnion: v1alpha2.ComponentUnion{
+			Plugin: &v1alpha2.PluginComponent{
+				ImportReference: v1alpha2.ImportReference{
+					RegistryUrl: url,
+				},
+			},
+		},
+	}
+}
+
+func TestValidateComponents(t *testing.T) {
+
+	volMounts := []v1alpha2.VolumeMount{
+		{
+			Name: "myvol",
+			Path: "/some/path/",
+		},
+	}
+
+	invalidVolMounts := []v1alpha2.VolumeMount{
+		{
+			Name: "myinvalidvol",
+		},
+		{
+			Name: "myinvalidvol2",
+		},
+	}
+
+	projectSourceEnv := []v1alpha2.EnvVar{
+		{
+			Name:  EnvProjectsSrc,
+			Value: "/some/path/",
+		},
+	}
+
+	projectsRootEnv := []v1alpha2.EnvVar{
+		{
+			Name:  EnvProjectsRoot,
+			Value: "/some/path/",
+		},
+	}
+
+	endpointUrl18080 := generateDummyEndpoint("url1", 8080)
+	endpointUrl18081 := generateDummyEndpoint("url1", 8081)
+	endpointUrl28080 := generateDummyEndpoint("url2", 8080)
+
+	invalidVolMountErr := ".*\nvolume mount myinvalidvol belonging to the container component.*\nvolume mount myinvalidvol2 belonging to the container component.*"
+	duplicateCmdErr := "duplicate key: component1"
```

Comment:
Maybe a copy-paste issue, but `Cmd` does not make sense in this context (maybe `duplicateComponentErr`?)